pkg/authentication/context.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package authentication
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+)
+
+type principalContextKey int
+
+var PrincipalContextKey principalContextKey
+
+type Principal struct {
+	Subject string `json:"sub"`
+}
+
+func NewPrincipalFromClaims(c ReadableClaims) (*Principal, error) {
+	a := new(Principal)
+	if err := c.Claims(a); err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+func PrincipalContext(ctx context.Context, principal *Principal) context.Context {
+	parent := ctx
+	if ctx == nil {
+		parent = context.Background()
+	}
+
+	return context.WithValue(parent, PrincipalContextKey, principal)
+}
+
+func PrincipalFromContext(ctx context.Context) *Principal {
+	if ctx == nil {
+		return nil
+	}
+
+	return ctx.Value(PrincipalContextKey).(*Principal)
+}
+
+type OAuth2Context struct {
+	client   *oauth2.Config
+	verifier TokenVerifier
+
+	tracer trace.Tracer
+	logger logging.LoggerInterface
+}
+
+func (o *OAuth2Context) LoginRedirect(w http.ResponseWriter, r *http.Request) {
+	_, span := o.tracer.Start(r.Context(), "authentication.OAuth2Context.LoginRedirect")
+	defer span.End()
+
+	additionalInfo, _ := json.Marshal(map[string]string{"referer": r.Referer(), "remote-address": r.RemoteAddr})
+	http.Redirect(
+		w, r,
+		o.client.AuthCodeURL(
+			base64.URLEncoding.EncodeToString(additionalInfo),
+			oauth2.SetAuthURLParam("audience", o.client.ClientID),
+		),
+		http.StatusFound,
+	)
+}
+
+func (o *OAuth2Context) RetrieveTokens(ctx context.Context, code string) (*oauth2.Token, error) {
+	_, span := o.tracer.Start(ctx, "authentication.OAuth2Context.RetrieveTokens")
+	defer span.End()
+
+	return o.client.Exchange(ctx, code)
+}
+
+func (o *OAuth2Context) RefreshToken(ctx context.Context, rawRefreshToken string) (*oauth2.Token, error) {
+	_, span := o.tracer.Start(ctx, "authentication.OAuth2Context.RefreshToken")
+	defer span.End()
+
+	return o.client.
+		TokenSource(ctx, &oauth2.Token{RefreshToken: rawRefreshToken}).
+		Token()
+}
+
+func (o *OAuth2Context) Verifier() TokenVerifier {
+	return o.verifier
+}
+
+func NewOAuth2Context(config *Config, getProvider OIDCProviderSupplier, tracer trace.Tracer, logger logging.LoggerInterface) *OAuth2Context {
+	o := new(OAuth2Context)
+	o.tracer = tracer
+	o.logger = logger
+
+	provider, err := getProvider(context.Background(), config.issuer)
+	if err != nil {
+		o.logger.Fatalf("Unable to fetch provider info, error: %v", err.Error())
+	}
+
+	var verifier TokenVerifier
```

Comment:
Just a **NONBLOCKING** question: do you think it's better to move this token verifier "factory" logic to `tokens.go`? It seems that `OAuth2Context` does not need to care about how to create an appropriate token verifier. It only needs to use the factory to get the right one.

Reply:
You are right and that was my original implementation, but since the factory process is so simple, I removed the factory classt and just made it part of the OAuth2Context, so it's one less object to crowd the implementation.
We can still extract it in case things get more complicated.